snuba/query/processors/arrayjoin_optimizer.py

```diff
@@ -0,0 +1,300 @@
+    Finds the top level conditions that include filter based on tags_key.
+    This is meant to be used to find the tag keys the query is filtering on.
+    We can only apply the arrayFilter optimization to tag keys conditions
+    that are not in OR with other columns. To simplify the problem, we only
+    consider those conditions that are included in the first level of the query:
+    [['tagskey' '=' 'a'],['col' '=' 'b'],['col2' '=' 'c']]  works
+    [[['tagskey' '=' 'a'], ['col2' '=' 'b']], ['tagskey' '=' 'c']] does not
+
+    If we encounter an OR condition we return None, which means we cannot
+    safely apply the optimization. Empty set means we did not find any
+    suitable tag key for optimization in this condition but that does
+    not disqualify the whole query in the way the OR condition does.
+    """
+    tag_keys_found = set()
+
+    conditions = get_first_level_conditions(condition)
+    for c in conditions:
+        if is_binary_condition(c, BooleanFunctions.OR):
+            return None
+
+        match = FunctionCall(
+            None,
+            String(ConditionFunctions.EQ),
+            (tags_key_pattern, Literal(None, Param("key", Any(str)))),
+        ).match(c)
+        if match is not None:
+            tag_keys_found.add(match.string("key"))
+
+        match = is_in_condition_pattern(tags_key_pattern).match(c)
+        if match is not None:
+            function = cast(FunctionCallExpr, match.expression("tuple"))
+            tag_keys_found |= {
+                l.value
+                for l in function.parameters
+                if isinstance(l, LiteralExpr) and isinstance(l.value, str)
+            }
+
+    return tag_keys_found
+
+
+def get_filtered_tag_keys(query: Query) -> Set[str]:
+    """
+    Identifies the tag names we can apply the arrayFilter optimization on.
+    Which means: if the tags_key column is in the select clause, there
+    are one or more top level AND condition on the tags_key column and
+    there is no OR condition in the query.
+    """
+    tags_key_found = any(
+        tags_key_pattern.match(f) is not None
+        for expression in query.get_selected_columns_from_ast() or []
+        for f in expression
+    )
+
+    if not tags_key_found:
+        return set()
+
+    ast_condition = query.get_condition_from_ast()
+    cond_tags_key = (
+        _get_tag_keys_in_condition(ast_condition)
+        if ast_condition is not None
+        else set()
+    )
+    if cond_tags_key is None:
+        # This means we found an OR. Cowardly we give up even though there could
+        # be cases where this condition is still optimizable.
+        return set()
+
+    ast_having = query.get_having_from_ast()
+    having_tags_key = (
+        _get_tag_keys_in_condition(ast_having) if ast_having is not None else set()
+    )
+    if having_tags_key is None:
+        # Same as above
+        return set()
+
+    return cond_tags_key | having_tags_key
+
+
+class ArrayjoinOptimizer(QueryProcessor):
```

Comment:
Turned it into ArrayJoinOptimizer. 
Do you mean making the name more descriptive? Because I have a very hard time describing the idea of joining two arrayjoin into one in a name.
Maybe ArrayJoinReducer ?

Reply:
I think using `Optimizer` in the name is fine but clarifying that this works on array joins of a specific form (mapping-style nested field with associated key, value pairs) would make this a bit more straightforward without having to read the docstring. Otherwise out of context this kind of implies it just works on array joins broadly which is not quite true.

Reply:
ahhh ok, sorry I misunderstood. Will revert and add some context.